### DISCUSSION TITLE
Relationship suppression not working when deployed to server

### DISCUSSION THREAD

**Discussion by kokoshneta** [filamentphp/filament#6378] [Help]
As described in [this previous comment](https://github.com/filamentphp/filament/discussions/2135#discussioncomment-5605294), I have a BookResource page with a DOI relationship that should be read-only. Existing DOI rows must never updated, and new DOI rows must never be inserted when editing a BookResource. As suggested, I used `->saveRelationshipsUsing(NULL)` to achieve this.

If there is an existing DOI record so the relationship is not null, it works as intended, and the DOI record is not updated.

But if there is no relationship, the call to `->saveRelationshipsUsing()` seems to be ignored: when I edit a BookResource which has no DOI relationship rows and hit “Save”, it attempts to insert a row in the DOI table (which then fails).

How can this be? Where might I look to find out what could be causing this particular method call to be ignored?

**Comment by kokoshneta** [comment]
I think I finally figured this out. I had been looking in the wrong places, trying to find some sort of `->createRelationshipRecordsUsing()` method, which doesn’t exist.

The answer was painfully simple: the [`->relationship()` method on layout components](https://filamentphp.com/docs/2.x/forms/layout#saving-data-to-relationships) accepts a second boolean/closure parameter titled `$saveCondition`. Inside the `->relationship()` method, this parameter is used in a call to `->saveRelationshipsUsing()`, which returns immediately if the `$saveCondition` evaluates to false.

So I just need to only return true in a creation context, and the layout component simply needs to be defined like this:

```php
Group::make([
  TextInput::make('doi')
    ->disabled()
    ->label('DOI')
])
  ->saveRelationshipsUsing(NULL)
  ->relationship('doi', fn ($context) => $context == 'create')
```

**Comment by danharrin** [comment]
I dont think you the `saveRelationshipsUsing(null)` if you pass a save condition

**Reply by kokoshneta** [reply]
I thought that might be the case, but I figured it won’t do any harm to keep it in.

**Comment by kokoshneta** [comment]
Unfortunately, it seems this wasn’t actually the fix after all. :-(

To specify, these are the conditions for the DOI field:

1. When creating a new book, _a new DOI should always be generated_
2. When updating an existing book, _the DOI table should always be unaffected_, meaning:
  a. if there is already a DOI record, it should be left as is
  b. if there is no DOI record, a new one should **not** be created

The solution above works for 1 and 2b – but in scenario 2a, saving the book **deletes** the DOI from the database, which should not happen.

It happens because of [this bit in `EntanglesStateWithSingularRelationship`](https://github.com/filamentphp/filament/blob/995326f6fcb41bc7a9f64957a489c0db372206d9/packages/forms/src/Components/Concerns/EntanglesStateWithSingularRelationship.php#L41):

```
if (! $component->evaluate($condition)) {
  $record?->delete();
   return;
}
```

Basically, it seems like the `$condition` parameter to the `->relationship()` method does not control whether the relationship is saved, but _whether it saves or deletes the relationship_.

Is there no way to prevent saving _without_ then causing deletion?

---

**Note:** This is still for Filament v2 – haven’t had the time to go through and upgrade to v3 yet.

**Comment by kokoshneta** [comment]
So it turns out that this _was_ actually the correct answer, right up until three days before I found it.

The deletion was **added** to the trait in [a commit on 10 May 2023](https://github.com/filamentphp/filament/commit/b72449002cc3f9cfee32572cb17f22a17d844b8e), apparently based on [this thread](https://github.com/filamentphp/filament/discussions/6440), though I don’t really see what deleting existing records – without confirmation, no less! – has to do with that thread.

@danharrin You added the feature last year – can you expand on why deletion was added if the save condition is not met? Was this requested in a thread that I can’t find? And is there no way to work around it?

**Reply by danharrin** [reply]
#6440 has a required field, if that is blank then the relationship is not saved. Adding deletion makes the behaviour consistent between Create/Edit, since saving the relationship without that field is not possible and results in no related record being created. We are unable to update the existing record without that field so it would result in unexpected behaviour of the data in the related record not reflecting the form after it is saved.

**Reply by kokoshneta** [reply]
@danharrin I don’t see how that’s consistent. Saving the relationship without the field is not possible, so if the _field is blank_ during creation, no related record is inserted. That is very far from the same as **actively deleting existing records with no user confirmation** if the _condition argument_ is empty or false, because the condition is not related to whether the field is empty.

As griggsk pointed out in #6440, it would be fair to delete the record if the field representing it had gone from being filled in to being empty (i.e., the user actively deletes the value). But the `$save` condition is settable based on anything at all. There are lots of scenarios where you may want to show a field in the form, but make it non-editable – two that immediately spring to mind are:

- My scenario here, where a record must be immutable and not editable after creation (condition is simply `FALSE`)
- A form where only super users have permission to update certain fields (condition is `fn() => $user->isSuperAdmin()`)

With the current implementation, **these scenarios are impossible**. Either the field is editable or the related record is deleted every time the field is submitted. To my mind, all three scenarios should be possible; for example, the signature could be changed to allow integers instead of a boolean and evaluating thus:

- if `$save > 0`, the related record is saved
- if `$save = 0`, the related record is left alone (not saved, not deleted)
- if `$save < 0`, the related record is deleted

That would make it look like this:

```
public function relationship(string $relationshipName, bool | int | Closure $condition = true): static {
  […]

  if (! $evaluatedCondition = $component->evaluate($condition)) {
    if ($evaluatedCondition < 0) $record?->delete();

    return;
  }

  […]
}
```

Or perhaps a third optional argument `$deleteIfNotSaving` could be added (default true to match current behaviour).

**Reply by danharrin** [reply]
The condition is related to whether or not the field is empty in the original case, and since it is false if the field is empty, we are unable to update the record. This is the intended and most common use case for `$condition` IIRC.

I suggest that you do not use the condition parameter. After `relationship()`, try `->when($form->getOperation() === 'edit', fn (Group $component) => $component->saveRelationshipsUsing(null)->saveRelationshipsBeforeChildrenUsing(null)`

**Reply by kokoshneta** [reply]
With the minor adjustment that `$form->getOperation()` isn’t available in v2 (I’m guessing it was introduced in v3), this does seem to work:

 ```php
Group::make()
  ->schema([
    TextInput::make('doi')
      ->disabled()
      ->default(fn ($context) => $context === 'create' ? DOI::generate() : NULL)
      ->label('DOI')
  ])
  ->relationship('doi')
  ->when(fn($context) => $context === 'edit', fn (Group $component) => $component
    ->saveRelationshipsUsing(null)
    ->saveRelationshipsBeforeChildrenUsing(null)
  )
;
```

**Reply by kokoshneta** [reply]
@danharrin Nope, turns out this didn’t work either after all. :-/

It works if the DOI field is filled out – it doesn’t attempt to update the existing row. But if the input field is empty, it attempts to insert a new row, which fails (since the field in the database is required) – exactly as described in the original post at the top.

At this point, I think I’ll just have to accept that there’s no way to accomplish this completely within Filament. So now I do the following instead:

I generate a DOI upon creation in the `Book` model:

```php
protected static function booted() {
  static::created(function (self $book) {
    $book->createDoi();
  });
}
```

And then I just never set the value in Filament and completely disable all types of saving for the relationship:

```php
Group::make()
 ->schema([
   TextInput::make('doi')
     ->disabled()
     ->label('DOI')
 ])
 ->relationship('doi')
 ->saveRelationshipsUsing(NULL)
 ->saveRelationshipsBeforeChildrenUsing(NULL)
;
```

It’s kind of a workaround, but at least this way it seems to work consistently both for creating and editing.

**Comment by abanghendri** [comment]
I use repeater to save many-to-many relationship and always got error `Call to a member function associate() on null ` at this line
```
//\Filament\Resources\Pages\EditRecord.php:134;
 $data = $this->form->getState(afterValidate: function () {
                $this->callHook('afterValidate');
 
                $this->callHook('beforeSave');
            });
```
What I did to solve this is just override save method on my EditResource to this:
```
 $data = $this->form->getState(shouldCallHooksBefore: false,, afterValidate: function () {
                $this->callHook('afterValidate');

                $this->callHook('beforeSave');
            });
```
just put `shouldCallHooksBefore: false` to `getState()` method, also remember to put `dehydrated()` method to repeater relationship, now you can modify how to save your relationship using handleRecordUpdate as mentioned [here](https://filamentphp.com/docs/3.x/panels/resources/editing-records#customizing-the-saving-process)
